Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -9,6 +16,346 @@ pub use num::NumCast;
+        if let Sequence::ArraySequence { values, .. } = self {
+            // The length of the elements
+            Rc::clone(values)
+        } else if let Sequence::ConcatSequence { length, .. } = self {
+            let mut array: Vec<T> = Vec::with_capacity(*length);
+            Sequence::<T>::append_recursive(&mut array, self);
+            Rc::new(array)
+        } else if let Sequence::LazySequence { boxed, is_string, ..  } = self {
+            let result = boxed.borrow().to_array();
+            // Put the value back into boxed
+            boxed.replace(Rc::clone(
+                &Sequence::<T>::new_array_sequence_is_string(&result, *is_string)));
+            result
+        } else {
+            panic!("This should never happen")
+        }
+    }
+
+    fn append_recursive(array: &mut Vec<T>, this: &Sequence<T>) {
+        match this {
+            Sequence::ArraySequence { values, .. } =>
+              // The length of the elements
+              for value in values.iter() {
+                array.push(value.clone());
+              },
+            Sequence::ConcatSequence { left, right, .. } =>
+              // Let's create an array of size length and fill it up recursively
+              {
+                Sequence::<T>::append_recursive(array, left);
+                Sequence::<T>::append_recursive(array, right);
+              }
+            Sequence::LazySequence { boxed, .. } =>
+            Sequence::<T>::append_recursive(array, boxed.borrow().as_ref()),
+        }
+    }
+
+    fn node_count(&self) -> usize {
+        match self {
+            Sequence::ArraySequence { node_count, .. } =>
+              // The length of the elements
+              *node_count,
+            Sequence::ConcatSequence { node_count, ..  } =>
+              *node_count,
+            Sequence::LazySequence { node_count, .. } =>
+              *node_count,
+        }
+    
+    }
+    /// Returns the cardinality of this [`Sequence<T>`].
+    // The cardinality returns the length of the sequence
+    fn cardinality(&self) -> SizeT {
+        match self {
+            Sequence::ArraySequence { values, .. } =>
+              // The length of the elements
+              values.len(),
+            Sequence::ConcatSequence { length, .. } =>
+              *length,
+            Sequence::LazySequence { length, .. } =>
+              *length,
+        }
+    }
+    fn select(&self, index: SizeT) -> T {
+        let array = self.to_array();
+        array[index].clone()
+    }
+    
+}
+
+// **************
+// Immutable maps
+// **************
+
+#[allow(dead_code)]
+struct Map<K, V> 
+  where K: Clone + Eq + std::hash::Hash, V: Clone
+{
+    data: Rc<Sequence<(K, V)>>,
+    // Any time we explicitly access this map, we index the data
+    cache: RefCell<Option<HashMap<K, V>>>
+}
+
+#[allow(dead_code)]
+impl <K, V> Map<K, V>
+  where K: Clone + Eq + std::hash::Hash, V: Clone
```

Comment:
One cross-cutting issue to keep in mind - there's an assumption that the image of all Dafny types implement `Hash`. What level of testing will support that? Do we have confidence eventually turning Rust on for all the uniform compiler tests will catch missing cases because the Rust won't compile?